// graphcool-cli-engine/dist/Output/index.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var StreamOutput_1 = require("./StreamOutput");
var chalk = require("chalk");
var styles = require("ansi-styles");
var ActionBase_1 = require("./actions/ActionBase");
var Prompter_1 = require("./Prompter");
var supports = require("supports-color");
var SpinnerAction_1 = require("./actions/SpinnerAction");
var SimpleAction_1 = require("./actions/SimpleAction");
var path = require("path");
var util = require("util");
var screen_1 = require("./actions/screen");
var ExitError_1 = require("../errors/ExitError");
var inquirer = require("graphcool-inquirer");
var migration_1 = require("./migration");
var treeify = require("treeify");
var dirTree = require("directory-tree");
var marked = require("marked");
var TerminalRenderer = require("marked-terminal");
marked.setOptions({
    renderer: new TerminalRenderer(),
});
exports.CustomColors = {
    // map gray -> dim because it's not solarized compatible
    supports: supports,
    gray: function (s) { return chalk.dim(s); },
    grey: function (s) { return chalk.dim(s); },
    attachment: function (s) { return chalk.cyan(s); },
    addon: function (s) { return chalk.yellow(s); },
    configVar: function (s) { return chalk.green(s); },
    release: function (s) { return chalk.blue.bold(s); },
    cmd: function (s) { return chalk.green.bold(s); },
    app: function (s) { return exports.CustomColors.graphcool("\u2B22 " + s); },
    graphcool: function (s) {
        if (!exports.CustomColors.supports) {
            return s;
        }
        var has256 = exports.CustomColors.supports.has256 ||
            (process.env.TERM || '').indexOf('256') !== -1;
        return has256
            ? '\u001b[38;5;104m' + s + styles.reset.open
            : chalk.magenta(s);
    },
};
function wrap(msg) {
    var linewrap = require('@heroku/linewrap');
    return linewrap(6, screen_1.errtermwidth, {
        skipScheme: 'ansi-color',
        skip: /^\$ .*$/,
    })(msg);
}
function bangify(msg, c) {
    var lines = msg.split('\n');
    for (var i = 0; i < lines.length; i++) {
        var line = lines[i];
        lines[i] = ' ' + c + line.substr(2, line.length);
    }
    return lines.join('\n');
}
function getErrorMessage(err) {
    var message;
    if (err.body) {
        // API error
        if (err.body.message) {
            message = util.inspect(err.body.message);
        }
        else if (err.body.error) {
            message = util.inspect(err.body.error);
        }
    }
    // Unhandled error
    if (err.message && err.code) {
        message = util.inspect(err.code) + ": " + err.message;
    }
    else if (err.message) {
        message = err.message;
    }
    return message || util.inspect(err);
}
var arrow = process.platform === 'win32' ? '!' : '▸';
var Output = (function () {
    function Output(config) {
        this.mock = true;
        this.config = config;
        this.mock = config.mock;
        this.stdout = new StreamOutput_1.default(process.stdout, this);
        this.stderr = new StreamOutput_1.default(process.stderr, this);
        this.action = ActionBase_1.shouldDisplaySpinner(this)
            ? new SpinnerAction_1.SpinnerAction(this)
            : new SimpleAction_1.SimpleAction(this);
        if (this.mock) {
            chalk.enabled = false;
            exports.CustomColors.supports = false;
        }
        this.prompter = new Prompter_1.default(this);
        this.prompt = inquirer.createPromptModule({ output: process.stdout });
        this.migration = new migration_1.MigrationPrinter(this);
    }
    Object.defineProperty(Output.prototype, "color", {
        get: function () {
            return new Proxy(chalk, {
                get: function (chalkProxy, name) {
                    if (exports.CustomColors[name]) {
                        return exports.CustomColors[name];
                    }
                    return chalkProxy[name];
                },
            });
        },
        enumerable: true,
        configurable: true
    });
    Output.prototype.log = function (data) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        (_a = this.stdout).log.apply(_a, [data].concat(args));
        var _a;
    };
    Output.prototype.getStyledJSON = function (obj) {
        var json = JSON.stringify(obj, null, 2);
        if (chalk.enabled) {
            var cardinal = require('cardinal');
            var theme = require('cardinal/themes/jq');
            return cardinal.highlight(json, { json: true, theme: theme });
        }
        else {
            return json;
        }
    };
    Output.prototype.done = function () {
        var rest = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            rest[_i] = arguments[_i];
        }
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.action.stop();
                return [2 /*return*/];
            });
        });
    };
    Output.prototype.debug = function (obj) {
        if (this.config.debug) {
            this.action.pause(function () { return console.log(obj); });
        }
    };
    Object.defineProperty(Output.prototype, "errlog", {
        get: function () {
            return path.join(this.config.cacheDir, 'error.log');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Output.prototype, "autoupdatelog", {
        get: function () {
            return path.join(this.config.cacheDir, 'autoupdate.log');
        },
        enumerable: true,
        configurable: true
    });
    Output.prototype.error = function (err, exitCode) {
        if (exitCode === void 0) { exitCode = 1; }
        if (this.mock && typeof err !== 'string' && exitCode !== false) {
            throw err;
        }
        try {
            if (typeof err === 'string') {
                err = new Error(err);
            }
            this.logError(err);
            if (this.action.task) {
                this.action.stop(this.color.bold.red('!'));
            }
            if (this.config.debug) {
                this.stderr.log(err.stack || util.inspect(err));
            }
            else {
                this.stderr.log(bangify(wrap(getErrorMessage(err)), this.color.red(arrow)));
                this.stderr.log('\nGet in touch if you need help: https://www.graph.cool/forum');
            }
        }
        catch (e) {
            console.error('error displaying error');
            console.error(e);
            console.error(err);
        }
        if (exitCode !== false) {
            this.exit(exitCode);
        }
    };
    Output.prototype.warn = function (err, prefix) {
        var _this = this;
        this.action.pause(function () {
            try {
                prefix = prefix ? prefix + " " : '';
                err = typeof err === 'string' ? new Error(err) : err;
                _this.logError(err);
                if (_this.config.debug) {
                    if (process.stderr.write("WARNING: " + prefix)) {
                        _this.stderr.log(err.stack || util.inspect(err));
                    }
                }
                else {
                    _this.stderr.log(bangify(wrap(prefix + _this.color.yellow(getErrorMessage(err))), _this.color.yellow(arrow)) + '\n');
                }
            }
            catch (e) {
                console.error('error displaying warning');
                console.error(e);
                console.error(err);
            }
        }, this.color.bold.yellow('!'));
    };
    Output.prototype.getPrettyModule = function (moduleName, type) {
        if (type === void 0) { type = 'error'; }
        var method = type === 'error' ? 'red' : 'yellow';
        return chalk[method]("[" + type.toUpperCase() + " in module: " + chalk.bold(moduleName) + "] ");
    };
    Output.prototype.logError = function (err) {
        StreamOutput_1.logToFile(util.inspect(err) + '\n', this.errlog);
    };
    Output.prototype.printMarkdown = function (markdown) {
        this.log(marked(markdown).split('\n').map(function (l) { return "   " + l; }).join('\n'));
    };
    Output.prototype.oldprompt = function (name, options) {
        return this.prompter.prompt(name, options);
    };
    Output.prototype.table = function (data, options) {
        var table = require('./table');
        return table(this, data, options);
    };
    Output.prototype.exit = function (code) {
        if (code === void 0) { code = 0; }
        if (this.config.debug) {
            console.error("Exiting with code: " + code);
        }
        if (this.mock) {
            throw new ExitError_1.default(code);
        }
        else {
            process.exit(code);
        }
    };
    Output.prototype.tree = function (dirPath, padding) {
        if (padding === void 0) { padding = false; }
        var tree = dirTree(dirPath);
        var convertedTree = treeConverter(tree);
        var printedTree = treeify.asTree(convertedTree, true);
        this.log(chalk.blue(printedTree.split('\n').map(function (l) { return (padding ? '   ' : '') + l; }).join('\n')));
    };
    return Output;
}());
exports.Output = Output;
function treeConverter(tree) {
    if (!tree.children) {
        return tree.name;
    }
    return tree.children.reduce(function (acc, curr) {
        if (!curr.children) {
            return __assign({}, acc, (_a = {}, _a[curr.name] = null, _a));
        }
        return __assign({}, acc, (_b = {}, _b[curr.name] = treeConverter(curr), _b));
        var _a, _b;
    }, {});
}
