// graphcool-cli-core/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var deploy_1 = require("./commands/deploy");
exports.Deploy = deploy_1.default;
var init_1 = require("./commands/init");
exports.Init = init_1.default;
var index_1 = require("./commands/auth/index");
exports.Auth = index_1.default;
var index_2 = require("./commands/info/index");
exports.Info = index_2.default;
var index_3 = require("./commands/playground/index");
exports.Playground = index_3.default;
var console_1 = require("./commands/console");
exports.Console = console_1.default;
var index_4 = require("./commands/list/index");
exports.List = index_4.default;
var index_5 = require("./commands/delete/index");
exports.Delete = index_5.default;
var index_6 = require("./commands/root-token/index");
exports.RootTokens = index_6.default;
var function_1 = require("./commands/logs/function");
exports.FunctionLogs = function_1.default;
var local_1 = require("./commands/invoke/local");
exports.InvokeLocal = local_1.default;
var pull_1 = require("./commands/local/pull");
exports.PullLocal = pull_1.default;
var stop_1 = require("./commands/local/stop");
exports.Stop = stop_1.default;
var up_1 = require("./commands/local/up");
exports.Up = up_1.default;
var restart_1 = require("./commands/local/restart");
exports.Restart = restart_1.default;
var account_1 = require("./commands/account");
exports.Account = account_1.default;
var eject_1 = require("./commands/local/eject");
exports.Eject = eject_1.default;
var add_template_1 = require("./commands/add-template/add-template");
exports.AddTemplate = add_template_1.default;
var ps_1 = require("./commands/local/ps");
exports.PsLocal = ps_1.default;
var reset_1 = require("./commands/reset/reset");
exports.Reset = reset_1.default;
var index_7 = require("./commands/clusters/index");
exports.Clusters = index_7.default;
var index_8 = require("./commands/import/index");
exports.Import = index_8.default;
var index_9 = require("./commands/export/index");
exports.Export = index_9.default;
exports.groups = [
    {
        key: 'general',
        name: 'General commands',
    },
    {
        key: 'data',
        name: 'Data workflows',
    },
    {
        key: 'local',
        name: 'Local development',
    },
    {
        key: 'platform',
        name: 'Platform',
    },
];
exports.topics = [
    { name: 'init', description: 'Create a new service', group: 'general' },
    {
        name: 'deploy',
        description: 'Deploy local service definition',
        group: 'general',
    },
    { name: 'login', description: 'Create account or login', group: 'platform' },
    { name: 'reset', description: 'Reset data of a service', group: 'data' },
    {
        name: 'playground',
        description: 'Opens the playground for the current service',
        group: 'general',
    },
    {
        name: 'info',
        description: 'Print service info (endpoints, clusters, ...) ',
        group: 'general',
    },
    { name: 'list', description: 'List all deployed services', group: 'general' },
    {
        name: 'root-token',
        description: 'Get the service root tokens',
        group: 'general',
    },
    { name: 'logs', description: 'Get logs of functions', group: 'general' },
    {
        name: 'delete',
        description: 'Delete a service',
        group: 'general',
    },
    {
        name: 'add-template',
        description: 'Add template a new template',
        group: 'general',
    },
    {
        name: 'invoke-local',
        description: 'Invokes a function locally',
        group: 'general',
    },
    {
        name: 'console',
        description: 'Opens the console for the current service',
        group: 'platform',
    },
    {
        name: 'account',
        description: 'Information about the current authenticated account',
        group: 'platform',
    },
    {
        name: 'local',
        description: 'Manage the local Graphcool version',
        group: 'local',
    },
    {
        name: 'clusters',
        description: 'Manage your private clusters',
        group: 'general',
    },
    {
        name: 'import',
        description: 'Import command',
        group: 'general',
    },
    {
        name: 'export',
        description: 'Export command',
        group: 'general',
    },
];
exports.commands = [
    deploy_1.default,
    init_1.default,
    index_1.default,
    index_2.default,
    index_3.default,
    console_1.default,
    index_4.default,
    index_5.default,
    index_6.default,
    function_1.default,
    local_1.default,
    pull_1.default,
    stop_1.default,
    up_1.default,
    restart_1.default,
    account_1.default,
    eject_1.default,
    add_template_1.default,
    ps_1.default,
    reset_1.default,
    index_7.default,
    index_8.default,
    index_9.default,
];
