// graphcool-cli-engine/dist/Config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var os = require("os");
var fs = require("fs-extra");
var cuid = require("cuid");
var debug = require('debug')('config');
var findUp = require("find-up");
var Config = (function () {
    function Config(options) {
        this.debug = Boolean(process.env.DEBUG && process.env.DEBUG.includes('*'));
        this.windows = false;
        this.bin = 'graphcool';
        this.mock = true;
        this.argv = process.argv.slice(1);
        this.commandsDir = path.join(__dirname, '../dist/commands');
        this.defaultCommand = 'help';
        this.userPlugins = false;
        this.version = '1.3.11';
        this.name = 'graphcool';
        this.pjson = {
            name: 'cli-engine',
            version: '0.0.0',
            dependencies: {},
            'cli-engine': {
                defaultCommand: 'help',
            },
        };
        this.root = path.join(__dirname, '..');
        this.warnings = [];
        this.authUIEndpoint = process.env.ENV === 'DEV'
            ? 'https://dev.console.graph.cool/cli/auth'
            : 'https://console.graph.cool/cli/auth';
        this.backendAddr = process.env.ENV === 'DEV'
            ? 'https://dev.api.graph.cool'
            : 'https://api.graph.cool';
        this.systemAPIEndpoint = process.env.ENV === 'DEV'
            ? 'https://dev.api.graph.cool/system'
            : 'https://api.graph.cool/system';
        this.authEndpoint = process.env.ENV === 'DEV'
            ? 'https://cli-auth-api.graph.cool/dev'
            : 'https://cli-auth-api.graph.cool/prod';
        this.docsEndpoint = process.env.ENV === 'DEV'
            ? 'https://dev.graph.cool/docs'
            : 'https://www.graph.cool/docs';
        this.statusEndpoint = 'https://crm.graph.cool/prod/status';
        /* tslint:disable-next-line */
        this.__cache = {};
        this.cwd = this.getCwd();
        this.home = this.getHome();
        this.setDotGraphcoolPath();
        this.setEnvPath();
        this.setDefinitionPaths();
        debug("dotGraphcoolPath after setting it", this.dotGraphcoolFilePath);
        this.setTokenIfExists();
        if (options) {
            this.readPackageJson(options);
        }
    }
    Config.prototype.setOutput = function (out) {
        this.out = out;
        this.warnings.forEach(function (warning) { return out.warn(warning); });
        this.warnings = [];
    };
    Config.prototype.setToken = function (token) {
        this.token = token;
    };
    Config.prototype.saveToken = function () {
        var json = JSON.stringify({ token: this.token }, null, 2);
        fs.writeFileSync(this.dotGraphcoolFilePath, json);
    };
    Object.defineProperty(Config.prototype, "arch", {
        get: function () {
            return os.arch() === 'ia32' ? 'x86' : os.arch();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Config.prototype, "platform", {
        get: function () {
            return os.platform() === 'win32' ? 'windows' : os.platform();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Config.prototype, "userAgent", {
        get: function () {
            return this.name + "/" + this.version + " (" + this.platform + "-" + this
                .arch + ") node-" + process.version;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Config.prototype, "dirname", {
        get: function () {
            return this.pjson['cli-engine'].dirname || this.bin;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Config.prototype, "cacheDir", {
        get: function () {
            return dir(this, 'cache', this.platform === 'darwin'
                ? path.join(this.home, 'Library', 'Caches')
                : null);
        },
        enumerable: true,
        configurable: true
    });
    Config.prototype.readPackageJson = function (options) {
        this.mock = options.mock;
        this.argv = options.argv || this.argv;
        if (options.root) {
            this.root = options.root;
            var pjsonPath = path.join(options.root, 'package.json');
            var pjson = fs.readJSONSync(pjsonPath);
            if (pjson && pjson['cli-engine']) {
                this.pjson = pjson;
                this.version = pjson.version;
            }
        }
    };
    Config.prototype.setEnvPath = function () {
        this.envPath = path.join(this.cwd, '.graphcoolrc');
        if (!fs.pathExistsSync(this.envPath)) {
            var found = findUp.sync('.graphcoolrc', { cwd: this.cwd });
            if (found) {
                // only use this if 1. it's not in the home dir (people of old cli versions may have this)
                // and 2. there MUST be a graphcool.yml file in the same folder, otherwise this makes no sense to use!
                var foundDir = path.dirname(found);
                if (foundDir === this.home) {
                    var file = fs.readFileSync(found);
                    var json = null;
                    try {
                        json = JSON.parse(file);
                    }
                    catch (e) {
                        //
                    }
                    if (json && json.token && (!this.dotGraphcoolFilePath || !fs.pathExistsSync(this.dotGraphcoolFilePath))) {
                        var newDotPath = path.join(this.home, '.graphcool');
                        fs.moveSync(found, newDotPath);
                        this.dotGraphcoolFilePath = newDotPath;
                    }
                    else {
                        this.warn("There is a .graphcoolrc file in your home directory (" + this.envPath + ").\nThis can still be an artifact of the old CLI version.\nTo prevent unwanted side effects, please remove it.");
                    }
                    return;
                }
                var ymlPath = path.join(foundDir, 'graphcool.yml');
                if (fs.pathExistsSync(ymlPath)) {
                    this.envPath = (found && path.dirname(found) !== this.home) ? found : this.envPath;
                }
            }
        }
    };
    Config.prototype.warn = function (msg) {
        this.warnings.push(msg);
    };
    Config.prototype.setDefinitionPaths = function () {
        var definitionPath = path.join(this.cwd, 'graphcool.yml');
        this.definitionDir = this.cwd;
        this.definitionPath = definitionPath;
        // if (fs.pathExistsSync(definitionPath)) {
        // } else {
        //   const found = findUp.sync('graphcool.yml', {cwd: this.cwd})
        //   this.definitionDir = found ? path.dirname(found) : this.cwd
        //   this.definitionPath = found || null
        // }
    };
    Config.prototype.setDotGraphcoolPath = function () {
        var dotGraphcoolCwd = path.join(this.cwd, '.graphcool');
        if (fs.pathExistsSync(dotGraphcoolCwd)) {
            this.dotGraphcoolFilePath = dotGraphcoolCwd;
        }
        else {
            var found = findUp.sync('.graphcool', { cwd: this.cwd });
            var dotGraphcoolHome = path.join(this.home, '.graphcool');
            // only take the find-up file, if it's "deeper" than the home dir
            this.dotGraphcoolFilePath = (found && (found.split('/').length > dotGraphcoolHome.split('/'))) ? found : dotGraphcoolHome;
        }
    };
    Config.prototype.setTokenIfExists = function () {
        if (process.env.NODE_ENV === 'test') {
            debug('taking graphcool test token');
            this.token = process.env.GRAPHCOOL_TEST_TOKEN;
        }
        else {
            if (fs.existsSync(this.dotGraphcoolFilePath)) {
                var configContent = fs.readFileSync(this.dotGraphcoolFilePath, 'utf-8');
                this.token = JSON.parse(configContent).token;
            }
        }
    };
    Config.prototype.getCwd = function () {
        // get cwd
        var cwd = process.cwd();
        if (process.env.NODE_ENV === 'test') {
            cwd = path.join(os.tmpdir(), cuid() + "/");
            fs.mkdirpSync(cwd);
            debug('cwd', cwd);
        }
        return cwd;
    };
    Config.prototype.getHome = function () {
        // get home
        var home = os.homedir() || os.tmpdir();
        if (process.env.NODE_ENV === 'test') {
            home = path.join(os.tmpdir(), cuid() + "/");
            fs.mkdirpSync(home);
            debug('home', home);
        }
        return home;
    };
    return Config;
}());
exports.Config = Config;
function dir(config, category, d) {
    var cacheKey = "dir:" + category;
    var cache = config.__cache[cacheKey];
    if (cache) {
        return cache;
    }
    d =
        d ||
            path.join(config.home, category === 'data' ? '.local/share' : '.' + category);
    if (config.windows) {
        d = process.env.LOCALAPPDATA || d;
    }
    d = process.env.XDG_DATA_HOME || d;
    d = path.join(d, config.dirname);
    fs.mkdirpSync(d);
    config.__cache[cacheKey] = d;
    return d;
}
